xmppserver/src/main/java/org/jivesoftware/openfire/ldap/LdapGroupProvider.java

```diff
@@ -124,7 +145,127 @@ private Group getGroupByDN(LdapName groupDN, Set<String> membersToIgnore) throws
+    
+                        paging=true;
+                        Log.debug("using range retrival for group: {}", relativeDN);
+                        break;
+                    }
+                }
+    
+                if (!paging)
+                {
+                    return processGroup(ctx, attrs, membersToIgnore);
+                }
+                else
+                {
+                    //retrieve first part with attributes that came from AD
+                    ArrayList<String> stdAttr=new ArrayList<String>();
+                    for (int n=0;n<standardAttributes.length;n++)
+                    {
+                        if (!standardAttributes[n].contains(manager.getGroupMemberField()))
+                        {
+                            stdAttr.add(standardAttributes[n]);
+                        }
+                    }
+    
+                    stdAttr.add(manager.getGroupMemberField()+";range=0-"+String.valueOf(rangehigh));
+                    Log.debug("first range will be 0-{}",rangehigh);
+    
+                    //reload Attributes as said in MSDN will prepare range retrival
+                    attrs = ctx.getAttributes(relativeDN, stdAttr.toArray(new String[stdAttr.size()]));
+    
+                    //get the first part of the group
+                    Group tmpGroup = processGroup(ctx, attrs, membersToIgnore);
+                    Collection<JID> admins = tmpGroup.getAdmins();
+                    String description = tmpGroup.getDescription();
+                    String name = tmpGroup.getName();
+                    ArrayList<JID> members = new ArrayList<JID>();
+                    members.addAll(tmpGroup.getMembers());
+    
+                    //now load the rest
+                    do
+                    {
+                        try
+                        {
+                            stdAttr.clear();
+                            for (int n=0;n<standardAttributes.length;n++)
+                            {
+                                if (!standardAttributes[n].contains(manager.getGroupMemberField()))
+                                {
+                                    stdAttr.add(standardAttributes[n]);
+                                }
+                            }
+                            //Add the next range rangehigh+1 to rangehigh+1+ldap.pagedResultsSize
+                            stdAttr.add(genNextRangePart(rangehigh));
+    
+                            attrs = ctx.getAttributes(relativeDN, stdAttr.toArray(new String[stdAttr.size()]));
+                            NamingEnumeration<? extends Attribute> inext = attrs.getAll();
+                            while (inext.hasMore())
+                            {
+                                Attribute attribute = inext.next();
+                                String key = attribute.getID().toLowerCase();
+                                if (key.startsWith(manager.getGroupMemberField())&&key.contains(";range="))
```

Comment:
Do I need anything specific in a test setup to hit this branch?

Reply:
You will need Active Directory with a Security Group and a minimum of 1500 members (maybe more)
We tested it like this:
- 1. Setup a Windows 2016 Server VM with Active Directory and created / added users with powershell in a loop
- 2. **change** Property **ldap.processBigGroups** (default: false) to **true** 
- 3. clear OF caches
- 4. open admin console in Openfire an try to load the group (select it here .../group-summary.jsp)

Active Directory will automatically add the "range" parameter on the first resultset with the range it gave back... 
Then we have to crawl through it to load the rest

Source:  https://docs.microsoft.com/de-de/previous-versions/windows/desktop/ldap/searching-using-range-retrieval?redirectedfrom=MSDN

Reply:
I'll take a look and see if that's got a comparable feature in OpenLDAP

Reply:
Although it'll also be useful to test that this doesn't break OpenLDAP without the feature too :)

Reply:
> Although it'll also be useful to test that this doesn't break OpenLDAP without the feature too :)

because of that you can enable/disable this feature completly (but it should only get active, if it finds the "range"-addition in the member-attributesname too) so in theory openldap should not be affected